src/multivariate/mvdiscretenonparametric.jl

```diff
@@ -0,0 +1,104 @@
+const MvDiscreteNonParametric{
+    T<:AbstractVector{<:Real},
+    P<:Real,
+    Ts<:AbstractVector{T},
+    Ps<:AbstractVector{P},
+} = GeneralDiscreteNonParametric{Multivariate,T,P,Ts,Ps}
+
+"""
+    MvDiscreteNonParametric(
+        support::AbstractVector,
+        p::AbstractVector{<:Real}=fill(inv(length(support)), length(support)),
+    )
+
+Construct a multivariate discrete nonparametric probability distribution with `support` and corresponding
+probabilities `p`. If the probability vector argument is not passed, then
+equal probability is assigned to each entry in the support.
+
+# Examples
+```julia
+# rows correspond to samples
+x = collect(eachrow(rand(10,2)))
+μ = MvDiscreteNonParametric(x)
+
+# columns correspond to samples
+y = collect(eachcol(rand(7,12)))
+ν = MvDiscreteNonParametric(y)
+```
+"""
+function MvDiscreteNonParametric(
+    support::AbstractArray{<:AbstractVector{<:Real}},
+    p::AbstractVector{<:Real} = fill(inv(length(support)), length(support)),
+)
+    return MvDiscreteNonParametric{eltype(support),eltype(p),typeof(support),typeof(p)}(
+        support,
+        p,
+    )
+end
+
+Base.eltype(::Type{<:MvDiscreteNonParametric{T}}) where {T} = T
+
+"""
+    support(d::MvDiscreteNonParametric)
+Get a sorted AbstractVector defining the support of `d`.
+"""
+support(d::MvDiscreteNonParametric) = d.support
+
+"""
+    probs(d::MvDiscreteNonParametric)
+Get the vector of probabilities associated with the support of `d`.
+"""
+probs(d::MvDiscreteNonParametric) = d.p
+
+
+Base.length(d::MvDiscreteNonParametric) = length(first(d.support))
+Base.size(d::MvDiscreteNonParametric) = (length(d), length(d.support))
```

Comment:
This is wrong and defined automatically based on `length`:

```suggestion
```

Reply:
Ok, done. But I'm still getting the erro :/, but instead it's:
```
StackOverflowError:

Stacktrace:
     [1] rand!(rng::Random._GLOBAL_RNG, s::MvDiscreteNonParametric{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, Float64, Vector{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}, Vector{Float64}}, x::SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true})
       @ Distributions ~/MEGA/EMAp/Distributions.jl/src/genericrand.jl:83
     [2] _rand!(rng::Random._GLOBAL_RNG, s::MvDiscreteNonParametric{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, Float64, Vector{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}, Vector{Float64}}, x::SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true})
       @ Distributions ~/MEGA/EMAp/Distributions.jl/src/genericrand.jl:117
     [3] rand!(rng::Random._GLOBAL_RNG, s::MvDiscreteNonParametric{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, Float64, Vector{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}, Vector{Float64}}, x::SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true})
       @ Distributions ~/MEGA/EMAp/Distributions.jl/src/genericrand.jl:91
--- the last 2 lines are repeated 14426 more times ---
 [28856] _rand!(rng::Random._GLOBAL_RNG, s::MvDiscreteNonParametric{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, Float64, Vector{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}, Vector{Float64}}, x::SubArray{Float64, 1, Matrix{Float64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true})
       @ Distributions ~/MEGA/EMAp/Distributions.jl/src/genericrand.jl:117
 [28857] rand!
       @ ~/MEGA/EMAp/Distributions.jl/src/genericrand.jl:91 [inlined]
 [28858] _rand!(rng::Random._GLOBAL_RNG, s::MvDiscreteNonParametric{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, Float64, Vector{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}, Vector{Float64}}, x::Matrix{Float64})
       @ Distributions ~/MEGA/EMAp/Distributions.jl/src/genericrand.jl:117
 [28859] rand!
       @ ~/MEGA/EMAp/Distributions.jl/src/genericrand.jl:108 [inlined]
 [28860] rand
       @ ~/MEGA/EMAp/Distributions.jl/src/multivariates.jl:21 [inlined]
 [28861] rand(s::MvDiscreteNonParametric{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, Float64, Vector{SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}, Vector{Float64}}, dims::Int64)
       @ Distributions ~/MEGA/EMAp/Distributions.jl/src/genericrand.jl:22
 [28862] top-level scope
       @ In[31]:1
 [28863] eval
       @ ./boot.jl:373 [inlined]
```

Reply:
I had added some changes. The `_rand!` was missing. I added it back, and it worked.